Peer/Apps/AppBuilder/CommandLineParser.cs

```diff
@@ -0,0 +1,102 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using CommandLine;
+using CommandLine.Text;
+using wimm.Secundatives;
+using Error = CommandLine.Error;
+
+namespace Peer.Apps.AppBuilder;
+
+public interface ICommandLineParser
+{
+    Result<Command, TextResult> Parse(string[] args);
+}
+
+public class CommandLineParser : ICommandLineParser
+{
+    private readonly List<IVerb> _verbs;
+
+    public CommandLineParser(IEnumerable<IVerb> verbs)
+    {
+        _verbs = verbs.ToList();
+    }
+
+    //CN -- Should just swap this over to a different result type:
+    // might be good to munge together the result type and the verb?
+    // ArgInfo { IVerb, ParseResult<T> }? 
+    // Actually it's really just a Result<IVerb, string>
+
+    public Result<Command, TextResult> Parse(string[] args)
+    {
+        //I'm just dealing with one layer of nesting right now. I'm lazy
+        var verbCount = args.Count(x => !x.StartsWith('-'));
+        var rootVerb = _verbs.FirstOrDefault(x => x.Name == args.FirstOrDefault());
+        var subs = rootVerb?.Subs.ToArray() ?? Array.Empty<IVerb>();
+        var parser = new Parser(config =>
+        {
+            config.AutoHelp = true;
+            config.AutoVersion = true;
+            config.IgnoreUnknownArguments = false;
+        });
+
+        //Cases:
+        // We don't find a verb that matches the first arg
+        // We find a verb that matches the first arg
+        //    There is more than one verb
+        //    There is a single verb
+        //        The verb has no handler
+
+        var state = new { verbCount, rootVerb, subCount = subs.Length };
+        // This is awful but no sleep makes for bad code
+        // Now that I'm looking at this handling this in a loop might have been
+        // better but I'm tired.
+        var parseResult = state switch
+        {
+            //If someone puts in "schboop" they should get the default help
+            { rootVerb: null } => Parse(parser, args, _verbs),
+            //If someone puts in a verb with children/no handler they should get the help for the children of the verb
+            { verbCount: 1, rootVerb.Handler: null } => Parse(parser, args[1..], subs),
+            //If someone puts in multiple words but the root verb has no subverbs they should get the default help
```

Comment:
I think the overall rule is that people should get the help for as much of the command as is valid. If someone puts in multiple non-option tokens, and the first one is a top level command but doesn't take subcommands then the other tokens are just arguments to the top level command. In that case we should show the help for that top level command I think.

Reply:
That's the general case of this (though I've got it described in terms of a single stage of nesting) I think this results in the thing you're asking for though